### PR TITLE
Adds a cheat to increment or decrement the current game year

### PR DESCRIPTION
**Pull request by LeftofZen**
https://github.com/OpenLoco/OpenLoco/issues/1277

### CODE REVIEW

**Comment by duncanspumpkin** [comment]
A tab just for time seems excessive. Surely this would be the only item in it. What other items might be grouped together in a tab. Perhaps weather, or Win Scenario? Maybe make the tab scenario cheats?

**Comment by LeftofZen** [comment]
@duncanspumpkin On one hand I agree, but on the other I am just following the existing structure of the cheats window. You'll notice the "Vehicle cheats" tab currently only has 2 options:
image

and so does the one I've added:
image

With regards to other items on this tab, I'd like to include more fine-grained control of the time, eg set the day, month, and year. I started this tab name as "Misc cheats", I'm happy to rename to "Scenario cheats" and expand in that direction if you prefer.

I'm still new to OpenLoco but I don't believe there is any 'weather' feature in the game like OpenRCT2? If there is I've missed it, in which case yes this tab could also include weather.

**Comment by LeftofZen** [comment]
I will note that I found no suitable icons to use as the tab icon so I opted for just the wrench.

**Comment by AaronVanGeffen** [comment]
Thanks for the patch! This should be a useful feature for many.

I think we should consider expanding the scope to set the full date, instead of just the year. The cheats window could display the current day, month, and year, together with [+][-] spinner buttons for each. Any thoughts on this?

> I'm still new to OpenLoco but I don't believe there is any 'weather' feature in the game like OpenRCT2? If there is I've missed it, in which case yes this tab could also include weather.

There is indeed no weather feature in Locomotion, nor OpenLoco.

**Comment by LeftofZen** [comment]
I can certainly expand the scope, but my thought was to add this change now, and expand on it in the next branch/PR. I prefer multiple small changes as opposed to one large change. If you think this change is too small, I will expand this to add the ability to change each of year/month/day.

**Comment by duncanspumpkin** [comment]
Yes sorry i was thinking about the snow feature and confusing it with openrct2. Making this a spinner shouldn't be too hard. I'm not saying we add all these things right now I'm more thinking about how things are named in the code.

As for images ignore that for now when we handle objects differently we can start adding more ui images.

**Comment by LeftofZen** [comment]
Cool cool, I'm down to make a spinner for each time field, I just figured getting my first commit in would be my priority for now. I'll add the spinners next time I have some spare coding time.

**Comment by LeftofZen** [comment]
I'll do this like in OpenRCT2, where the date is done on the "Financial cheats" tab:
image

File: src/OpenLoco/Windows/Cheats.cpp
Comment (LeftofZen) [inline src/OpenLoco/Windows/Cheats.cpp:387]: can replace all of these with maths if we care enough

**Comment by LeftofZen** [comment]
This is ready for review if a maintainer would like to have a look at it.
image

File: src/OpenLoco/Date.cpp
Comment (LeftofZen) [inline src/OpenLoco/Date.cpp:147]: if date.year < 1800 this won't work

**Comment by LeftofZen** [comment]
@duncanspumpkin How can I get this reviewed? Seems like it needs a maintainer to look at it?

**Comment by duncanspumpkin** [comment]
> @duncanspumpkin How can I get this reviewed? Seems like it needs a maintainer to look at it?

Currently I'm away will be back to normal service next week

**Comment by LeftofZen** [comment]
bump

File: src/OpenLoco/Windows/Cheats.cpp
Comment (duncanspumpkin) [inline src/OpenLoco/Windows/Cheats.cpp:402]: Ive created a Math::Bound::sub and add for this exact situation.

File: src/OpenLoco/GameCommands/Cheat.cpp
Comment (duncanspumpkin) [inline src/OpenLoco/GameCommands/Cheat.cpp:181]: Avoid using C style casts use a static_cast instead.

File: src/OpenLoco/Date.cpp
Comment (duncanspumpkin) [inline src/OpenLoco/Date.cpp:143]: Future refactor to move these constants into header.

File: src/OpenLoco/Date.cpp
Comment (duncanspumpkin) [inline src/OpenLoco/Date.cpp:141]: Days? Think it should be plural.

**Review by duncanspumpkin — APPROVED** [review]
This all seems sensible. Can you address the comments and then this is good to go.

File: src/OpenLoco/Windows/Cheats.cpp
Comment (LeftofZen) [reply src/OpenLoco/Windows/Cheats.cpp:402]: Your `sub()` and `add()` functions need a third parameter for the bound for them to be useful here, because they hard code `numeric_limits<T>::min()` (and `max()`), which I only use once here, in `case Widx::year_step_increase:`. I'm happy to add that if you agree

Comment (duncanspumpkin) [reply src/OpenLoco/Windows/Cheats.cpp:403]: Hmm good point. Can you remove the c style casts at least. Modifying the bound function can be done in a follow on

**Comment by duncanspumpkin** [comment]
Ive merged the other PR first so could you rebase this one, remove the c style casts and this is good to go.

**Comment by LeftofZen** [comment]
(don't merge yet, need to test)

**Comment by LeftofZen** [comment]
looks like day is off-by-one
image

**Comment by LeftofZen** [comment]
Fixed off-by-one but I found another bug - if you put the date to a high number, like 31 for December, then move the month back to November, you can set it to November 31. This doesn't have any serious effect though; if you run the cheat it will set the date to December 1.
Let me know if you want me to iterate this, or if not then feel free to merge

**Comment by duncanspumpkin** [comment]
Could you add the changelog entry then we can merge this.

**Comment by LeftofZen** [comment]
Done

**Pull request merged**
